storage2/src/snapshot.rs

```diff
@@ -9,38 +9,38 @@ use tracing::Span;
 
 use crate::state::StateRead;
 
-/// Snapshots maintain a point-in-time view of the underlying storage, suitable
-/// for read-only access by multiple threads, i.e. RPC calls.
+mod rocks_wrapper;
+use rocks_wrapper::RocksDbSnapshot;
+
+/// A snapshot of the underlying storage at a specific state version, suitable
+/// for read-only access by multiple threads, e.g., RPC calls.
 ///
-/// This is implemented as a wrapper around a [RocksDB snapshot](https://github.com/facebook/rocksdb/wiki/Snapshot)
-/// with an associated JMT version number for the snapshot.
+/// Snapshots are cheap to create and clone.  Internally, they're implemented as
+/// a wrapper around a [RocksDB
+/// snapshot](https://github.com/facebook/rocksdb/wiki/Snapshot) with a pinned
+/// JMT version number for the snapshot.
 #[derive(Clone)]
-pub(crate) struct Snapshot(pub(crate) Inner);
+pub struct Snapshot(Arc<Inner>);
 
 // We don't want to expose the `TreeReader` implementation outside of this crate.
-#[derive(Clone)]
-pub(crate) struct Inner {
-    // TODO: the `'static` lifetime is a temporary hack and we'll need to find a workaround separately (tracked in #1512)
-    rocksdb_snapshot: Arc<rocksdb::Snapshot<'static>>,
-    db: &'static rocksdb::DB,
-    jmt_version: jmt::Version,
+struct Inner {
+    snapshot: RocksDbSnapshot,
+    version: jmt::Version,
+    // Used to retrieve column family handles.
+    db: Arc<rocksdb::DB>,
```

Comment:
It's slightly dissatisfying that we have to keep the `db` handle around for column family lookups-by-string, rather than storing `jmt_cf`, ... etc, `ColumnFamily` instances. The reason we have to do this is that the `ColumnFamily` type from `rocksdb` is a wrapper around an FFI'd column family pointer, so it's neither `Copy` nor `Clone`, and the `DB` only gives a borrowed `&ColumnFamily` from a hashmap that it stores internally.  So avoiding the hashmap would mean building something similar to the `RocksDbSnapshot` "internal Arc" wrapper but for column families, which doesn't seem worthwhile.